File: lib/three.rb
require 'opal'

require 'js'
require 'native'

module Three
    def self.const_missing(name)
        ThreeClass.new(name)
    end

    class ThreeClass
        def initialize(name)
            @name = name

            @class = $$[:THREE][name]
        end

        def new(*args)
            Native(JS.new(@class, *args.map(&:to_n)))
        end

        def method_missing(name, *args, &block)
            $$[:THREE][@name][name]
        end
    end

    class STLLoader
        def self.new(*args)
            Native(JS.new($$[:STLLoader], *args.map(&:to_n)))
        end
    end

    class STLExporter
        def self.new(*args)
            Native(JS.new($$[:STLExporter], *args.map(&:to_n)))
        end
    end

    class OrbitControls
        def self.new(*args)
            Native(JS.new($$[:OrbitControls], *args.map(&:to_n)))
        end
    end
end

File: node/three.js
import * as THREE from 'three';

import { STLLoader } from "three/examples/jsm/loaders/STLLoader.js";
import { STLExporter } from "three/examples/jsm/exporters/STLExporter.js";

import { OrbitControls } from "three/examples/jsm/controls/OrbitControls.js";

window.THREE = THREE;

window.STLLoader = STLLoader;
window.STLExporter = STLExporter;

STLLoader.prototype.parseBase64 = function(base64String) {
    let binaryString = atob(base64String);

    let uint8Array = Uint8Array.from(binaryString, c => c.charCodeAt(0));

    return new STLLoader().parse(uint8Array.buffer);
};

window.OrbitControls = OrbitControls;